subsys/shell/modules/devmem_service.c

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <shell/shell.h>
+#include <stdlib.h>
+
+static int memory_read(const struct shell *sh, mem_addr_t addr, uint8_t width)
+{
+	uint64_t value;
+	int err = 0;
+
+	switch (width) {
+	case 8:
+		value = sys_read8(addr);
+		break;
+	case 16:
+		value = sys_read16(addr);
+		break;
+	case 32:
+		value = sys_read32(addr);
+		break;
+	default:
+		shell_fprintf(sh, SHELL_NORMAL, "Incorrect data width\n");
+		err = -EINVAL;
+		break;
+	}
+
+	if (err == 0) {
+		shell_fprintf(sh, SHELL_NORMAL, "Read value 0x%lx\n", value);
+	}
+
+	return err;
+}
+
+static int memory_write(const struct shell *sh, mem_addr_t addr,
+			uint8_t width, uint64_t value)
+{
+	int err = 0;
+
+	switch (width) {
+	case 8:
+		sys_write8(value, addr);
+		break;
+	case 16:
+		sys_write16(value, addr);
+		break;
+	case 32:
+		sys_write32(value, addr);
+		break;
+	default:
+		shell_fprintf(sh, SHELL_NORMAL, "Incorrect data width\n");
+		err = -EINVAL;
+		break;
+	}
+
+	return err;
+}
+
+/* The syntax of the command is similar to busybox's devmem */
+static int cmd_devmem(const struct shell *sh, size_t argc, char **argv)
+{
+	mem_addr_t phys_addr, addr;
+	uint32_t value = 0;
+	uint8_t width;
+
+	if (argc < 2 || argc > 4) {
+		return -EINVAL;
+	}
+
+	phys_addr = strtol(argv[1], NULL, 16);
+
+#if defined(CONFIG_MMU) || defined(CONFIG_PCIE)
+	device_map((mm_reg_t *)&addr, phys_addr, 0x100, K_MEM_CACHE_NONE);
+
+	shell_fprintf(sh, SHELL_NORMAL, "Mapped 0x%lx to 0x%lx\n",
```

Comment:
You can consider using shell_print(shell, "Mapped 0x%lx to 0x%lx", phys_addr, addr); to fit it in one line

Reply:
Done, thanks.